lib/databases/timescale/amm/amm.go

```diff
@@ -0,0 +1,263 @@
+	commonApps "github.com/fluidity-money/fluidity-app/common/ethereum/applications"
+	"github.com/fluidity-money/fluidity-app/lib/log"
+	"github.com/fluidity-money/fluidity-app/lib/timescale"
+	"github.com/fluidity-money/fluidity-app/lib/types/applications"
+	"github.com/fluidity-money/fluidity-app/lib/types/ethereum"
+	"github.com/fluidity-money/fluidity-app/lib/types/misc"
+	"github.com/fluidity-money/fluidity-app/lib/types/network"
+	token_details "github.com/fluidity-money/fluidity-app/lib/types/token-details"
+	"github.com/fluidity-money/fluidity-app/lib/types/worker"
+)
+
+const (
+	// Context to use for logging
+	Context = `TIMESCALE/AMM`
+
+	// TableAmmPositions to store amm positions in
+	TableAmmPositions = `amm_positions`
+
+	// TablePendingLpRewards to track unpaid LP rewards
+	TablePendingLpRewards = `pending_lp_rewards`
+)
+
+func InsertAmmPosition(mint amm.PositionMint) {
+	timescaleClient := timescale.Client()
+
+	var (
+		id    = mint.Id
+		lower = mint.Lower
+		upper = mint.Upper
+		pool  = mint.Pool
+
+		statementText string
+	)
+
+	statementText = fmt.Sprintf(
+		`INSERT INTO %s (
+            position_id,
+            tick_lower,
+            tick_upper,
+			pool,
+            liquidity
+		) VALUES (
+			$1,
+			$2,
+			$3,
+            $4,
+			$5
+		);`,
+		TableAmmPositions,
+	)
+
+	_, err := timescaleClient.Exec(
+		statementText,
+		id,
+		lower,
+		upper,
+		pool,
+		misc.BigIntFromInt64(0),
+	)
+
+	if err != nil {
+		log.Fatal(func(k *log.Log) {
+			k.Context = Context
+			k.Message = "Failed to insert a new position!"
+			k.Payload = err
+		})
+	}
+}
+
+func UpdateAmmPosition(update amm.PositionUpdate) {
+	timescaleClient := timescale.Client()
+
+	var (
+		id    = update.Id
+		delta = update.Delta
+
+		statementText string
+	)
+
+	statementText = fmt.Sprintf(
+		`UPDATE %s
+            SET liquidity = liquidity + $1
+            WHERE position_id = $2
+		;`,
+		TableAmmPositions,
+	)
+
+	_, err := timescaleClient.Exec(
+		statementText,
+		delta,
+		id,
+	)
+
+	if err != nil {
+		log.Fatal(func(k *log.Log) {
+			k.Context = Context
+			k.Message = "Failed to update a position's liquidity!"
+			k.Payload = err
+		})
+	}
+}
+
+func HandleAmmWinnings(win worker.EthereumWinnerAnnouncement, tokenDetails map[applications.UtilityName]token_details.TokenDetails) {
+	if win.Application != commonApps.ApplicationSeawaterAmm || win.Decorator == nil {
+		return
+	}
+
+	var (
+		ammPrices = win.Decorator.ApplicationData.AmmPrices
+
+		firstToken  = ammPrices.FirstToken
+		firstTick   = ammPrices.FirstTick
+		secondToken = ammPrices.SecondToken
+		secondTick  = ammPrices.SecondTick
+		wins        = win.ToWinAmount
+	)
+
+	if firstToken == ethereum.AddressFromString("") {
+		return
+	}
+
+	if secondToken == ethereum.AddressFromString("") {
+		// swap1
+		addAmmWins(win, wins, tokenDetails, firstToken, firstTick, false)
+		return
+	}
+
+	// swap2
+	addAmmWins(win, wins, tokenDetails, firstToken, firstTick, true)
```

Comment:
This could be a problem with the lack of atomicity in the insertion, so we need to log the beginning of this insertion, the ending, and have a precommit step with a random snowflake, then have it set right after these changes to insert "done"... we'll fix the spooler behaviour later